### PR TITLE
DM-25164: Add diaForceSources to diaPipe loader and to alert packets

### PR DESCRIPTION
**Pull request by morriscb**

### CODE REVIEW

File: python/lsst/ap/association/diaPipe.py [hunk 1]
```diff
@@ -239,9 +241,14 @@ def run(self, diaSourceCat, diffIm, exposure, ccdExposureIdBits):
             exposure.getInfo().getVisitInfo().getDate().toPython())
         self.apdb.storeDiaForcedSources(diaForcedSources)
         if self.config.doPackageAlerts:
+            diaForcedSources.set_index("diaObjectId", drop=False, inplace=True)
+            if len(loaderResult.diaForcedSources) > 1:
+                diaForcedSources = pd.concat([diaForcedSources,
+                                              loaderResult.diaForcedSources])
```
Comment (spenczar) [inline python/lsst/ap/association/diaPipe.py:247]: I'm kind of surprised by the order here, I'd have guessed that you need to `set_index` _after_ doing the `concat` with `loaderResult`. Maybe this is fine, I don't really know, just mentioning it.

Comment (spenczar) [reply python/lsst/ap/association/diaPipe.py:247]: Oh yeah, I see the same index is set in `loadDiaForcedSources`, which I guess is where this value comes from, right?

File: python/lsst/ap/association/loadDiaCatalogs.py [hunk 2]
```diff
@@ -192,6 +197,35 @@ def loadDiaSources(self, diaObjects, pixelRanges, dateTime, apdb):
...
         return diaSources.replace(to_replace=[None], value=np.nan)
 
+    @pipeBase.timeMethod
+    def loadDiaForcedSources(self, diaObjects, dateTime, apdb):
+        """Load DiaObjects from the Apdb based on their HTM location.
+
+        Parameters
+        ----------
+        pixelRanges : `tuple` [`int`]
+            Ranges of pixel values that cover region of interest.
```
Comment (spenczar) [inline python/lsst/ap/association/loadDiaCatalogs.py:207]: This looks out of sync with the function params

File: tests/test_packageAlerts.py [hunk 3]
```diff
@@ -167,7 +167,45 @@ def makeDiaSources(nSources, diaObjectIds, exposure):
...
+    Parameters
+    ----------
+    nSources : `int`
+        Number of sources to create.
+    diaObjectIds : `numpy.ndarray`
+        Integer Ids of diaobjects to "associate" with the DiaSources.
+    exposure : `lsst.afw.image.Exposure`
+        Exposure to create sources over.
+    pixelator : `lsst.sphgeom.HtmPixelization`
+        Object to compute spatial indicies from.
```
Comment (spenczar) [inline tests/test_packageAlerts.py:182]: This isn't in the function signature, maybe out of date

File: python/lsst/ap/association/loadDiaCatalogs.py [hunk 4]
```diff
@@ -192,6 +197,35 @@ def loadDiaSources(self, diaObjects, pixelRanges, dateTime, apdb):
...
+            DiaObjects loaded from the Apdb that are within the area defined
+            by ``pixelRanges``.
+        """
+        if len(diaObjects) == 0:
+            # If no diaObjects are available return an empty DataFrame with
+            # the the column used for indexing later in AssociationTask.
+            diaForcedSources = pd.DataFrame(columns=["diaObjectId"])
+        else:
+            diaForcedSources = apdb.getDiaForcedSources(
+                diaObjects.loc[:, "diaObjectId"],
```
Comment (spenczar) [inline python/lsst/ap/association/loadDiaCatalogs.py:223]: Should the object IDs be trimmed down to just unique values? Maybe it's already guaranteed to be unique, but what happens if there are object IDs that appear multiple times, are forced sources duplicated in the result here?

File: python/lsst/ap/association/diaPipe.py (hunk 1 above)
Comment (morriscb) [reply python/lsst/ap/association/diaPipe.py:247]: Yeah, what you're seeing here was an attempt to get around the fact that when joined, dataframes can do weird things with the types. Now that you mention it, I'm not sure the data ends up sorted here. I'll switch back to an appended with a specific "sort" to make this more explicit.

File: python/lsst/ap/association/loadDiaCatalogs.py (hunk 4 above)
Comment (morriscb) [reply python/lsst/ap/association/loadDiaCatalogs.py:223]: So, the daiObjectIds are duplicated for sure but the row values are unique so I don't think there will be any duplicated data here. I believe pandas groups together the duplicated values, but what I can do is specifically add the diaForcedSourceId here to make sure this is what is happening and get an ordering for the diaForcedSources in the alert packet.

File: python/lsst/ap/association/diaPipe.py (hunk 1 above)
Comment (morriscb) [reply python/lsst/ap/association/diaPipe.py:247]: I moved the diaForcedSources indexing into the diaForcedSource task to keep it more compartmentalized.

File: python/lsst/ap/association/loadDiaCatalogs.py (hunk 2 above)
Comment (morriscb) [reply python/lsst/ap/association/loadDiaCatalogs.py:207]: Sorry! Copy paste bites me again.

**Comment by morriscb** [comment]
Hey @spenczar, could you take a quick look, please, and see if I have responded to your review adequately. Thanks!

**Comment by spenczar** [comment]
@morriscb I think it all makes sense, but I don't see any commits pushed, just to be clear.

**Comment by morriscb** [comment]
Oops, sorry. Just pushed now.

File: python/lsst/ap/association/loadDiaCatalogs.py [hunk 5]
```diff
@@ -192,6 +197,41 @@ def loadDiaSources(self, diaObjects, pixelRanges, dateTime, apdb):
...
+        dataTime : `datetime.datetime`
+            Time of the current visit
+        apdb : `lsst.dax.apdb.Apdb`
+            Database connection object to load from.
+
+        Returns
+        -------
+        diaObjects : `pandas.DataFrame`
+            DiaObjects loaded from the Apdb that are within the area defined
+            by ``pixelRanges``.
```
Comment (spenczar) [inline python/lsst/ap/association/loadDiaCatalogs.py:218]: `pixelRanges` is out of date

File: python/lsst/ap/association/loadDiaCatalogs.py [hunk 6]
```diff
@@ -192,6 +197,41 @@ def loadDiaSources(self, diaObjects, pixelRanges, dateTime, apdb):
...
+    @pipeBase.timeMethod
+    def loadDiaForcedSources(self, diaObjects, dateTime, apdb):
+        """Load DiaObjects from the Apdb based on their HTM location.
+
+        Parameters
+        ----------
+        diaObjects : `pandas.DataFrame`
+            DiaObjects loaded from the Apdb that are within the area defined
+            by ``pixelRanges``.
+        dataTime : `datetime.datetime`
```
Comment (spenczar) [inline python/lsst/ap/association/loadDiaCatalogs.py:209]: typo: `dataTime` -> `dateTime`

**Pull request merged**